docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,602 @@
+  parameters supported and can contain some practical examples about how to use it.
+- `columns`: Information about the event that the gadget provides. Each element represents a columns attribute as defined in pkg/columns/columninfo.go.
+
+## Gadget Implementation
+
+A gadget is composed by eBPF programs and maps that are compiled into an eBPF object. According to
+the names of the maps and programs, and the types used, Inspektor Gadget is able to understand how
+to parse, enrich and print information from the gadget.
+
+Currently the built-in gadgets have a strong categorization, however in this new containerized
+gadgets approach that categorization is softer, each gadget defines how it behaves according to the
+maps / types they use and it's possible that the same gadget implements different behaviors, like
+tracer and topper, etc.
+
+### Maps
+
+#### `RingBuf` and `PerfEventArray` with `print_` Prefix (a.k.a tracers)
+
+These maps indicate that the gadget produces a stream of events. Inspektor Gadget uses the BTF
+information to format and print the events.
+
+```c
+// taken from gadgets/trace_open.bpf.c
+struct event {
+	__u64 timestamp;
+	__u32 pid;
+	__u32 uid;
+	__u32 gid;
+	mnt_ns_id_t mntns_id;
+	int ret;
+	int flags;
+	__u16 mode;
+	__u8 comm[TASK_COMM_LEN];
+	__u8 fname[NAME_MAX];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__type(value, struct event);
+} print_events SEC(".maps");
+```
+
+#### `HashMap` with `stats_` Prefix (a.k.a toppers)
+
+These maps are used to implement toppers, i.e. gadgets that print a list of elements sorted by
+specific parameters, like TCP connections by sent bytes.
+
+Inspektor Gadget automatically reads and sorts those maps according to the configuration provided by the user:
+- Interval: How often to print the map contents
+- Sort by: Field within the event type to sort by
+
+#### BPF Iterators (a.k.a snapshotters)
+
+Programs of type `iter/` are automatically loaded and attached by Inspektor Gadget, then they are
+triggered and its output is parsed according to the BTF information provided by it.
+
+In order to have BTF information available, the programs have to define an unused variable with the
+type they output in the iterator:
+
+```c
+const volatile event_t myiterator_output = {};
+
+const struct myevent *gadget_iter_type __attribute__((unused));
+
+SEC("iter/task_file")
+int myiterator(struct bpf_iter__task_file *ctx)
+```
+
+#### `HashMap` with `hist_` Prefix (a.k.a profilers)
+
+These maps are use to implement profilers: gadgets that output a histogram. The histogram collection
+starts when the gadget is run, and it's printed when the gadgets stops.
+
+TODO1: It's possible to also print the histogram periodically, however it's not explored yet.
+
+TODO2: There is probably some overlap with histogram metrics support for Prometheus. It's very
+likely that the same gadget can be used for both purposes.
+
+#### Prometheus
```

Comment:
I don't understand why this is a special category and not an export option for various other gadgets. What I'd imagined is that certian types of gadgets will be able to configure support for Prometheus. Then in the config file we could specify how that works and the user could then make use of that at runtime if that's what they wanted. Optionally, they could expose it as an API or imply have it output to standard out.

Reply:
It's not about having another category, but about describing how a gadget should implement that support. All this section describes the different features gadgets can implement, Prometheus is just another of those features. In order to enable prometheus export, a given gadget will need to define some maps in a specific format, like `count_`, `histogram_`, et. (TBD), then Inspektor Gadget will expose those prometheus metrics automatically.